Description: Add a gsettings key to use VTE "alternate scrolling" and default it true.
 There's no UI - a change from previously. Advanced users can change this in gsettings.
Author: Iain Lane <iain.lane@canonical.com>
Forwarded: not-needed (vte API is not upstream yet)
Bug-Upstream: https://bugzilla.gnome.org/show_bug.cgi?id=518405

Index: b/src/migration.c
===================================================================
--- a/src/migration.c
+++ b/src/migration.c
@@ -58,6 +58,7 @@
 #define GCONF_PROFILES_PREFIX  GCONF_PREFIX "/profiles"
 
 #define KEY_ALLOW_BOLD "allow_bold"
+#define KEY_ALTERNATE_SCREEN_SCROLL "alternate_screen_scroll"
 #define KEY_BACKGROUND_COLOR "background_color"
 #define KEY_BACKGROUND_DARKNESS "background_darkness"
 #define KEY_BACKGROUND_IMAGE_FILE "background_image"
@@ -425,6 +426,9 @@
   migrate_bool (client, path, KEY_SCROLL_ON_KEYSTROKE,
                 settings, TERMINAL_PROFILE_SCROLL_ON_KEYSTROKE_KEY,
                 FALSE);
+  migrate_bool (client, path, KEY_ALTERNATE_SCREEN_SCROLL,
+                settings, TERMINAL_PROFILE_ALTERNATE_SCREEN_SCROLL_KEY,
+                FALSE);
   migrate_bool (client, path, KEY_SCROLL_ON_OUTPUT,
                 settings, TERMINAL_PROFILE_SCROLL_ON_OUTPUT_KEY,
                 FALSE);
Index: b/src/org.gnome.Terminal.gschema.xml.in
===================================================================
--- a/src/org.gnome.Terminal.gschema.xml.in
+++ b/src/org.gnome.Terminal.gschema.xml.in
@@ -182,6 +182,11 @@
       <_summary>Whether to scroll to the bottom when there's new output</_summary>
       <_description>If true, whenever there's new output the terminal will scroll to the bottom.</_description>
     </key>
+    <key name="alternate-screen-scroll" type="b">
+      <default>true</default>
+      <_summary>Whether to use VTE "alternate scrolling"</_summary>
+      <_description>If true, "alternate scrolling" support will be enabled, allowing the mouse wheel to be used to scroll some kinds of windows, e.g. terminal pagers such as "less".</_description>
+    </key>
     <key name="exit-action" enum="org.gnome.Terminal.ExitAction">
       <default>'close'</default>
       <_summary>What to do with the terminal when the child command exits</_summary>
Index: b/src/terminal-schemas.h
===================================================================
--- a/src/terminal-schemas.h
+++ b/src/terminal-schemas.h
@@ -31,6 +31,7 @@
 #define TERMINAL_PROFILES_LIST_SCHEMA   "org.gnome.Terminal.ProfilesList"
 
 #define TERMINAL_PROFILE_ALLOW_BOLD_KEY                 "allow-bold"
+#define TERMINAL_PROFILE_ALTERNATE_SCREEN_SCROLL_KEY    "alternate-screen-scroll"
 #define TERMINAL_PROFILE_AUDIBLE_BELL_KEY               "audible-bell"
 #define TERMINAL_PROFILE_BACKGROUND_COLOR_KEY           "background-color"
 #define TERMINAL_PROFILE_BACKSPACE_BINDING_KEY          "backspace-binding"
Index: b/src/terminal-screen.c
===================================================================
--- a/src/terminal-screen.c
+++ b/src/terminal-screen.c
@@ -901,6 +901,10 @@
   if (!prop_name || prop_name == I_(TERMINAL_PROFILE_SCROLL_ON_OUTPUT_KEY))
     vte_terminal_set_scroll_on_output (vte_terminal,
                                        g_settings_get_boolean (profile, TERMINAL_PROFILE_SCROLL_ON_OUTPUT_KEY));
+  if (!prop_name || prop_name == I_(TERMINAL_PROFILE_ALTERNATE_SCREEN_SCROLL_KEY))
+    vte_terminal_set_alternate_screen_scroll (vte_terminal,
+                                              g_settings_get_boolean (profile,
+                                                TERMINAL_PROFILE_ALTERNATE_SCREEN_SCROLL_KEY));
   if (!prop_name ||
       prop_name == I_(TERMINAL_PROFILE_SCROLLBACK_LINES_KEY) ||
       prop_name == I_(TERMINAL_PROFILE_SCROLLBACK_UNLIMITED_KEY))
